docs/design_notes/lift.md

```diff
@@ -0,0 +1,340 @@
+# Lifted transformation
+
+⚠️ Advanced topic ⚠️
+
+This design note explains the underlying implementation of `flax.linen.transform`, which enables JAX transformations inside `Module`.
+
+
+## Introduction
+
+JAX uses a functional API meaning that it only guarantees correct behavior when using functions without side effects.
+Typically, these side effects are the result of mutating an object that lives outside the function.
+
+The functional paradigm has some advantages like the ability to explicitly reason about state and stochasticity.
+The function output only changes when an input argument changes.
+Therefore, a function is guaranteed to behave deterministically.
+
+But pure functions offer another big advantage to JAX: specifically, they enable functional transformations.
+For example `jax.vmap(f)` will vectorize a function `f`.
+Because `f` cannot have side effects the vectorized/parallel version of `f` is well-defined. To see why we need this restriction, consider what happens if `f` would increment a counter or draw a random number.
+Would `f` draw the same or a different random number for each item in the vector?
+Would each item in the batch have its own counter or is the counter shared among the items?
+And in what order is the counter incremented if `f` is computed in parallel?
+The answer to all these questions is "it depends".
+The behavior is ambiguous and the functional constraint elegantly avoids this problem.
+
+Flax introduces a safe way to have limited randomness and stateful variables in a JAX compatible form.
+The reason why the state in Flax is not problematic is because it is local: inside a Flax `Module` there are variables and PRNG sequences,
+but on the outside there are only JAX Arrays and PRNG keys.
+
+For most use cases, Flax is used to define models in a stateful way.
+Because a `Module` behaves like a pure function externally, we can fully utilize JAX with all of its transformations.
+There are, however, cases when we want to have the best of both worlds by using transformations and `Module` together.
+This design note explains how we extend JAX's functional transformation to work on `Module`s that have internal state and randomness.
+
+
+## Functionalization
+
+Before we jump into the details let's consider a simple example where we would like to use `vmap` inside a `Module`.
+
+First, we define a simple MLP without any transformations:
+
+```python
+import jax
+from jax import random, NumPy as jnp
```

Comment:
it's `numpy` (lowercase)
anyways, usually we would say `import jax.numpy as jnp` on a separate line...

(also if you use `jax.random` instead of `random` then I don't think any import is needed in the example since they're pretty obvious)

Reply:
oops